routing/router.go

```diff
@@ -941,45 +944,79 @@ func (r *ChannelRouter) SendPayment(payment *LightningPayment) ([32]byte, *Route
 	)
 
 	var (
-		sendError error
-		preImage  [32]byte
+		preImage [32]byte
 	)
 
-	// TODO(roasbeef): consult KSP cache before dispatching
+	if len(payment.Routes) == 0 {
+		// TODO(roasbeef): consult KSP cache before dispatching
 
-	// Before attempting to perform a series of graph traversals to find
-	// the k-shortest paths to the destination, we'll first consult our
-	// path cache
-	rt := newRouteTuple(payment.Amount, payment.Target)
+		// Before attempting to perform a series of graph traversals to find
+		// the k-shortest paths to the destination, we'll first consult our
+		// path cache
+		rt := newRouteTuple(payment.Amount, payment.Target)
 
-	r.routeCacheMtx.RLock()
-	routes, ok := r.routeCache[rt]
-	r.routeCacheMtx.RUnlock()
+		r.routeCacheMtx.RLock()
+		routes, ok := r.routeCache[rt]
+		r.routeCacheMtx.RUnlock()
 
-	// If we don't have a set of routes cached, we'll query the graph for a
-	// set of potential routes to the destination node that can support our
-	// payment amount. If no such routes can be found then an error will be
-	// returned.
-	if !ok {
-		freshRoutes, err := r.FindRoutes(payment.Target, payment.Amount)
-		if err != nil {
-			return preImage, nil, err
+		// If we don't have a set of routes cached, we'll query the graph for a
+		// set of potential routes to the destination node that can support our
+		// payment amount. If no such routes can be found then an error will be
+		// returned.
+		if !ok {
+			freshRoutes, err := r.FindRoutes(payment.Target, payment.Amount)
+			if err != nil {
+				return preImage, nil, err
+			}
+
+			// Populate the cache with this set of fresh routes so we can
+			// reuse them in the future.
+			r.routeCacheMtx.Lock()
+			r.routeCache[rt] = freshRoutes
+			r.routeCacheMtx.Unlock()
+
+			routes = freshRoutes
 		}
 
-		// Populate the cache with this set of fresh routes so we can
-		// reuse them in the future.
-		r.routeCacheMtx.Lock()
-		r.routeCache[rt] = freshRoutes
-		r.routeCacheMtx.Unlock()
+		// Update *LightningPayment with routes we just have found
+		payment.Routes = routes
 
-		routes = freshRoutes
+		// Attempt to send this payment through the routes we just have found
+		return r.SendToRoute(payment)
+	} else if len(payment.Routes) > 0 {
+		return r.SendToRoute(payment)
 	}
 
-	// For each eligible path, we'll attempt to successfully send our
+	return [32]byte{}, nil, fmt.Errorf("unable to send payment")
+}
+
+// SendToRoute attempts to send a payment  through the passed
```

Comment:
All lines should be wrapped to 80 characters on a best effort basis as is done nearly uniformly elsewhere in the project.